### ISSUE TITLE
Get coordinates of centroids

### ISSUE THREAD

**Issue by delanoe** [alpmestan/kmeans-vector#6]
Hello,
is it possible to get the coordinates only (and not the clusters for instances) ?
Cheers,

**Comment by alpmestan** [comment]
Hello, (ou bien bonjour =P)

Centroids are not returned by anything exposed to the user at the moment. I would take a patch that does it though if that's useful and if you feel like putting one together? My newborn eagerly consumes all my free time these days =)

**Comment by delanoe** [comment]
Hello, (oui bonjour pourquoi pas ;)
I was wondering if the coordinates were computed already (then I just need to find the way to give it back) or coordinates are not computed (a quick look on the code let me think it is possible) so I have to calculate it.

Cheers,
(Congratulations for your newborn.)

**Comment by alpmestan** [comment]
They are computed by `centroidsOf`, which is called in `kmeansStep`: https://github.com/alpmestan/kmeans-vector/blob/master/Math/KMeans.hs#L154

You basically want to write a variant of `kmeansStep` and `kmeansWith` that doesn't "forget" about the centroids but return them along with the clusters.

**Comment by delanoe** [comment]
Hello,
I try to add type signature to anonymous functions inside kmeansWith in order to write the variant. I have an idea how I could do it but I encounter some type issues.

So, for instance, if I add type signature to the `go` function I get the error below (and If i change the type into the expected type I got the same error almost). Do you know why ? Is that the reason why you commented it in the code ? Thanks for your help.

   Couldn't match type ‘a1’ with ‘a’
      ‘a1’ is a rigid type variable bound by
           the type signature for go :: Clusters a1 -> Clusters a1
           at /kmeans-vector/Math/KMeans.hs:147:11
      ‘a’ is a rigid type variable bound by
          the type signature for
            kmeansWith :: Monad m =>
                          (Int -> [a] -> m (Clusters a))
                          -> (a -> V.Vector Double)
                          -> Distance
                          -> Int
                          -> [a]
                          -> m (Clusters a)
          at /kmeans-vector/Math/KMeans.hs:137:15
    Expected type: G.Vector (Cluster a)
      Actual type: Clusters a1
    Relevant bindings include
      pgroups :: Clusters a1
        (bound at /kmeans-vector/Math/KMeans.hs:148:8)
      go :: Clusters a1 -> Clusters a1
        (bound at /kmeans-vector/Math/KMeans.hs:148:5)
      kmeansStep :: G.Vector (Cluster a) -> G.Vector (Cluster a)
        (bound at /kmeans-vector/Math/KMeans.hs:154:5)
      centroidsOf :: G.Vector (Cluster a) -> G.Vector (V.Vector Double)
        (bound at /kmeans-vector/Math/KMeans.hs:163:5)
      pairToClosestCentroid :: G.Vector (V.Vector Double)
                               -> a -> (Int, a)
        (bound at /kmeans-vector/Math/KMeans.hs:174:5)
      pgroupsEqualUnder :: G.Vector (Cluster a)
                           -> G.Vector (Cluster a) -> Bool
        (bound at /kmeans-vector/Math/KMeans.hs:178:5)
      initF :: Int -> [a] -> m (Clusters a)
        (bound at /kmeans-vector/Math/KMeans.hs:144:12)
      (Some bindings suppressed; use -fmax-relevant-binds=N or -fno-max-relevant-binds)
    In the first argument of ‘kmeansStep’, namely ‘pgroups’
    In the expression: kmeansStep pgroups
Failed, modules loaded: none.

**Comment by alpmestan** [comment]
I think you just need to turn the `ScopedTypeVariable` language extension on, at the top of the module?

``` haskell
{-# LANGUAGE ScopedTypeVariables #-}
module [...] where
```

I think I left the type sigs commented out just in order to avoid having to turn that extension on (although it's harmless), but I don't remember, it's been a while =)

**Comment by delanoe** [comment]
Here are the centroids kept in a tuple:

https://github.com/delanoe/kmeans-vector/blob/anoe/master/Math/KMeans.hs#L201

Maybe the tuple is not the right data structure.

I am thinking about something like a data structure keeping both Centroid and Clusters.

Do you have any advice ?

**Comment by alpmestan** [comment]
Yeah I guess we could make `Cluster` a data type with an additional field that points to the centroid of the said cluster.
